app/command_giphy.go

```diff
@@ -0,0 +1,255 @@
+				Height   string `json:"height"`
+				Size     string `json:"size"`
+				Webp     string `json:"webp"`
+				WebpSize string `json:"webp_size"`
+			} `json:"fixed_width_downsampled"`
+			DownsizedMedium struct {
+				URL    string `json:"url"`
+				Width  string `json:"width"`
+				Height string `json:"height"`
+				Size   string `json:"size"`
+			} `json:"downsized_medium"`
+			Original struct {
+				URL      string `json:"url"`
+				Width    string `json:"width"`
+				Height   string `json:"height"`
+				Size     string `json:"size"`
+				Frames   string `json:"frames"`
+				Mp4      string `json:"mp4"`
+				Mp4Size  string `json:"mp4_size"`
+				Webp     string `json:"webp"`
+				WebpSize string `json:"webp_size"`
+			} `json:"original"`
+			FixedHeight struct {
+				URL      string `json:"url"`
+				Width    string `json:"width"`
+				Height   string `json:"height"`
+				Size     string `json:"size"`
+				Mp4      string `json:"mp4"`
+				Mp4Size  string `json:"mp4_size"`
+				Webp     string `json:"webp"`
+				WebpSize string `json:"webp_size"`
+			} `json:"fixed_height"`
+			Looping struct {
+				Mp4     string `json:"mp4"`
+				Mp4Size string `json:"mp4_size"`
+			} `json:"looping"`
+			OriginalMp4 struct {
+				Width   string `json:"width"`
+				Height  string `json:"height"`
+				Mp4     string `json:"mp4"`
+				Mp4Size string `json:"mp4_size"`
+			} `json:"original_mp4"`
+			PreviewGif struct {
+				URL    string `json:"url"`
+				Width  string `json:"width"`
+				Height string `json:"height"`
+				Size   string `json:"size"`
+			} `json:"preview_gif"`
+		} `json:"images"`
+	} `json:"data"`
+	Meta struct {
+		Status     int    `json:"status"`
+		Msg        string `json:"msg"`
+		ResponseID string `json:"response_id"`
+	} `json:"meta"`
+}
+
+func (me *GiphyProvider) DoCommand(args *model.CommandArgs, message string) *model.CommandResponse {
+	// TODO: this is a developer API key that cannot be used in production
+	// we have to upgrade it to a production key before releasing this feature
+	// see https://developers.giphy.com/docs/
+	giphyApiKey := "*****"
+
+	escapedQuery := url.PathEscape(message)
+	giphyUrl := fmt.Sprintf("https://api.giphy.com/v1/gifs/translate?api_key=%s&s=%s", giphyApiKey, escapedQuery)
+
+	resp, err := http.Get(giphyUrl)
+	if err != nil {
+		l4g.Error("An error occurred while querying Giphy, err=%v", err)
+	}
+	defer resp.Body.Close()
+
+	var giphyResponse giphyTranslateResponse
+	if bytes, err := ioutil.ReadAll(resp.Body); err != nil {
+		l4g.Error("An error occurred while reading the Giphy response body, err=%v", err)
+	} else if err := json.Unmarshal(bytes, &giphyResponse); err != nil {
+		l4g.Error("An error occurred while deserializing the Giphy response, err=%v", err)
+	}
+
+	post := &model.Post{
+		ChannelId: args.ChannelId,
+		RootId:    args.RootId,
+		ParentId:  args.ParentId,
+		UserId:    args.UserId,
+		Message:   fmt.Sprintf("%v\n%v", message, giphyResponse.Data.Images.FixedHeight.URL),
```

Comment:
This is my hacky way of moving this content to the UI. @hmhealey mentioned that there was some way to embed context information in a post. If I could send a rich JSON object instead of a simple text message, I'd be able to do something a little more satisfying on the UI side.

Reply:
You can pretty much throw arbitrary stuff into "props".